### ISSUE TITLE
[Improvement]

### ISSUE THREAD

**Issue by Erlix322** [lukechilds/hyperdex#262]
Could you add a cancel Button in the Exchange view?
Would be great to cancel a pending order directly in there.

---

HyperDEX 0.1.0-alpha.1
Electron 2.0.0
win32 10.0.16299
Locale: de

**Comment by sindresorhus** [comment]
Thanks for the suggestion. We plan to add this.

**Comment by lukechilds** [comment]
Currently this functionality isn't very useful due to the fact that you can only cancel pending swaps. Normally a swap will either be matched fairly quickly or never match.

Cancelling swaps that never match obviously doesn't add any real usability. Cancelling a swap before it's matched is difficult as it gets matched so fast. And cancelling a swap after it's matched would be useful for the user and is probably what most users expect to be able to do but isn't allowed by `marketmaker`.

I raised this before with James and asked if we could cancel matched swaps but he was strongly opposed to it. Is there any chance you would reconsider @jl777?

**Comment by jl777** [comment]
we could slow the ordermatch down a lot, but that seems not so good.

alternatively, changing the protocol would be required. What do you suggest to be done? Alice would need to force bob to reclaim the payment and deposit. I am not against it, it just seems quite impractical

**Comment by lukechilds** [comment]
>we could slow the ordermatch down a lot, but that seems not so good.

No, fast matching is good.

>alternatively, changing the protocol would be required. What do you suggest to be done? Alice would need to force bob to reclaim the payment and deposit. I am not against it, it just seems quite impractical

@jl777 couldn't you just stop Alice from responding to Bob anymore? So as far as Bob's concerned the swap failed due to network/communication issues. Then the swap will unwind/reclaim using the normal procedure.

Obviously id there's any way to instantly cancel and reclaim funds that's great, but I'm guessing that's not possible due to the timelock.

If so we will display a clear message to the user when they click cancel. We'll let them know that they'll lose their dexfee and it may take up to 4 hours for their funds to be reclaimed. We'll require them to confirm they're ok with that before actually cancelling.

**Comment by jl777** [comment]
user can effectively do this just by exiting, but I see how it is good to do it your way. I would need to add external signalling to the main swap threads. If we can limit to one cancel at a time, that would enable a global variable based method, it would then generate a "swapcancelled" message so you would know when you can cancel another one

**Comment by jl777** [comment]
but you want to cancel swaps that are already in progress, dont you?
there is no limit to the number of those, other than the practical limit of how many you can get going before they start completing

**Comment by lukechilds** [comment]
>but you want to cancel swaps that are already in progress, dont you?

Yeah. I was assuming the mm would respond with a success after it's cancelled. So mm wouldn't process a new cancel until the first one finished.

But after re-reading your message with regards to the socket message. It sounds like mm would respond to the request before it's finished cancelling and then we get a message later notifying us when it's successfully cancelled.

>If we can limit to one cancel at a time, that would enable a global variable based method, it would then generate a "swapcancelled" message so you would know when you can cancel another one

If I understand this correctly you want to have a single global variable that signals if a swap is cancelling so issuing multiple calls would mess this up. Is there a reason it needs to be implemented like that. Could you add something like an `isCancelling` property to the specific trade that is cancelling? Or have list of cancelling trades rather than a single value? That way we could cancel multiple trades at once.

Or is that not possible?

**Comment by jl777** [comment]
once an atomic swap starts, it launches a dedicated thread for that swap
to send messages to a specific swap, then I need to implement a messaging system from the main mm process to all the threads. Not impossible, but:

UUID_being_cancelled = uuid;

above is a much simpler way to implement it. if you update the UUID_being_cancelled before it actually is noticed and cancelled by the dedicated thread, it just wouldnt be cancelled

**Comment by lukechilds** [comment]
What bout a global list of UUIDs rather than just a reference to a single UUID?

Would that get around the message passing complications without imposing any user restrictions?

**Comment by jl777** [comment]
good idea, i guess this global list wont ever get too big, so the api would accept an array of uuid strings and abort at the protocol level, ie. same effect as pulling the internet cable.

so whatever array is passed in will set the global cancel list.

yes that would work, cant guarantee at what stage the swap is stopped and the usually timelocks need to expire to complete via normal swapstatus(requestid,quoteid), but it will prevent any new tx from being sent out

**Comment by lukechilds** [comment]
I was thinking the list would just be internal to mm.

So we pass in a single UUID with the API, it gets added to the global list, and removed once it's completed.

**Comment by jl777** [comment]
sorry, there are unknown number of threads running. getting it to accurately know when all have been processed would require adding some global threads tracking, which doesnt exist now.

I wanted to avoid all this complexity internally as mm1.0 is at the limits of having new requirements added to it after it is all coded and debugged.

That is why I suggested a singled global uuid. Now this is become some global thread manangement... maybe we make it a 2.0 issue.

also the timeout for electrum connect seems to have destabilized trades, investigating that

**Comment by lukechilds** [comment]
@Erlix322 This will be added as part of #481. It works much better with GTC orders, however the mm cancel behaviour is still slightly buggy.
